presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java

```diff
@@ -165,11 +170,36 @@ public synchronized void addLocation(URI location)
                 scheduler,
                 pageBufferClientCallbackExecutor);
         allClients.put(location, client);
+        taskIdToURI.put(sourceTaskId, location);
```

Comment:
If we are not sure about the TODO mentioned above, maybe we should check if `sourceTaskId` is already in the list and if so, fail loudly. Or standing on the safer side, e.g. using `Multimap`.